### PR TITLE
All Product Reviews: Moderation

### LINKED ISSUE #1391: Product Reviews: Review Moderation
Parent project: #1324

Implement product review moderation. This includes:
- Trash
- Spam
- Approve/unapprove
- UNDO support

### PR DESCRIPTION
**Pull request by AmandaRiu**
Fixes #1391 by added the logic necessary for moderating product reviews. It includes the following functionality:
- full moderation support
- full undo support
- address prior bugs with review moderation (see comments below)

## Moderation + reopen

The current product version of review moderation is a bit on the buggy side. Previously, if you swiped the UNDO snackbar closed and then immediately opened the product review, you'd see the old status and then the new status would appear as soon as the request to update the status completed on the backend. This was a timing issue. To get around this, I created a new flow using a `ProductReviewModerationRequest` that also contains the property `RequestStatus`. This property is important because the ViewModel uses it to update the UI of the status of the request. I use a base class of `DelayedUndoRequest` so this design can be reused for similarly awkward scenarios where the users request to complete an action isn't _actually_ processed until the **UNDO** snackbar is dismissed. Now the UI will not clear the value of `pendingModerationNewStatus` - which is used to populate the status if the review is opened into the detail view - before the request to change the status has successfully completed.

## Moderation + undo

# Recommended Test Scenarios

- [x] Mark a review as approved, dismiss UNDO snackbar, open review and verify correct status displayed.
- [x] Pull to refresh the list of reviews, open the same review from above and verify the status is still correct. This validates the request was sent to the API and changed the status successfully.
- [x] Mark a review as approved and while the UNDO snackbar is visible, open the review and verify correct status displayed.
- [x] Mark a review as approved and click the UNDO button to revert the change. Open the review and verify correct status displayed.
- [x] Mark a review as spam - verify the correct review is removed the list. While the UNDO snackbar is visible, pull to refresh the list. Verify the previously spammed review does not reappear in the list.
- [x] Mark a review as spam - click UNDO. Verify the correct review is added back to the list.
- [x] Mark an **unread** review as trash - click UNDO. Verify the correct review is added back to the list.
- [x] Find a section with only **one** review - mark that review as trash. Verify both the section and the review are removed. Click UNDO - verify both the section and the review are added back to the view.

### CODE REVIEW

**Comment by AmandaRiu** [comment]
This PR is ready for review! 🎆

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/DelayedUndoRequest.kt [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package com.woocommerce.android.model
+
+abstract class DelayedUndoRequest(var requestStatus: RequestStatus) {
```
Comment (nbradbury) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/model/DelayedUndoRequest.kt]: This class would benefit from a comment explaining its purpose.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewDetailFragment.kt [hunk 2]
```diff
@@ -187,9 +207,54 @@ class ReviewDetailFragment : BaseFragment() {
...
+
+        // Use the status override if present,else new status
+        when (val newStatus = navArgs.tempStatus?.let { ProductReviewStatus.fromString(it) } ?: status) {
+            APPROVED -> review_approve.isChecked = true
+            HOLD -> review_approve.isChecked = false
+            else -> WooLog.w(REVIEWS, "Unable to process Review with a status of $newStatus")
+        }
+
+        // Configure the moderate button
+        with(review_approve) {
```
Comment (nbradbury) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewDetailFragment.kt]: np: these three `with` blocks aren't necessary.

**Comment by nbradbury** [comment]
The review detail closes if you attempt to moderate when there's no connection. In this situation, should the review detail show the message about no connection and then stay active?

Screenshot_1569351499

File: WooCommerce/src/main/res/layout/fragment_review_detail.xml [hunk 3]
```diff
@@ -151,35 +151,38 @@
                 android:id="@+id/review_approve"
                 android:layout_width="wrap_content"
                 android:layout_height="wrap_content"
+                android:layout_marginBottom="8dp"
```
Comment (nbradbury) [inline WooCommerce/src/main/res/layout/fragment_review_detail.xml]: We could use `@dimen/margin_medium` in this layout instead of hard-coding 8dp.

**Comment by nbradbury** [comment]
@AmandaRiu I'm done with my first round!

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/DelayedUndoRequest.kt (hunk 1 above)
Comment (AmandaRiu) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/model/DelayedUndoRequest.kt]: After reviewing this further I decided I didn't need a base class specific to a "DelayedUndo" request type. Did some consolidation and renaming in 0d8cffb20477cf8cc01e47f21ebfa0fa0c264138

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewDetailFragment.kt (hunk 2 above)
Comment (AmandaRiu) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewDetailFragment.kt]: Great catch! These were a holdover from a time when there was more work being done in these blocks. Removed in 6e20299e4cbd5a6e6ae97ae3e850bbaf48dda9d5

File: WooCommerce/src/main/res/layout/fragment_review_detail.xml (hunk 3 above)
Comment (AmandaRiu) [reply WooCommerce/src/main/res/layout/fragment_review_detail.xml]: Fixed in 5f44e09b521aa78ecff1b4b5ac8df6a121980934

**Comment by AmandaRiu** [comment]
@nbradbury code comments addressed directly. I also now show the "offline" error in the Review detail view in c777beb1f0c12366a9e674de161241602e4e07ba ...which actually works much better - so thank you for pointing that out 👍 Ready for another round!

**Comment by nbradbury** [comment]
Looks great! :shipit:

**Pull request merged**